rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs

```diff
@@ -0,0 +1,1089 @@
+        self.rewind();
+    }
+
+    // Rewinds the cursor to target immediately after the last indexed snapshot.
+    fn rewind(&mut self) {
+        self.current_indexing_snapshot = self.last_indexed_snapshot.next_target();
+    }
+}
+
+#[async_trait]
+impl<T: Sequenced + Debug> ContractSyncCursor<T> for ForwardSequenceAwareSyncCursor<T> {
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        // TODO: Fix ETA calculation
+        let eta = Duration::from_secs(0);
+        if let Some(range) = self.get_next_range().await? {
+            Ok((CursorAction::Query(range), eta))
+        } else {
+            // TODO: Define the sleep time from interval flag
+            Ok((CursorAction::Sleep(Duration::from_secs(5)), eta))
+        }
+    }
+
+    // TODO: revisit to establish a better heuristic for cursor / indexing health
+    fn latest_block(&self) -> u32 {
+        self.current_indexing_snapshot.at_block
+    }
+
+    /// Updates the cursor with the logs that were found in the range.
+    ///
+    /// Inconsistencies in the logs are not considered errors, instead they're handled by rewinding the cursor
+    /// to retry ranges.
+    ///
+    /// ## logs
+    /// The logs to ingest. If any logs are duplicated or their sequence is lower than the current indexing snapshot,
+    /// they are filtered out.
+    ///
+    /// If Sequence mode:
+    /// - The sequences of the logs must exactly match the range.
+    /// - If there are any gaps, the cursor rewinds and the range will be retried.
+    ///
+    /// If Block mode:
+    /// - Empty logs are allowed, but no gaps are allowed. The logs must build upon the last indexed snapshot.
+    /// - If there are any gaps, the cursor rewinds to the last indexed snapshot, and ranges will be retried.
+    /// - If the target block is reached and the target sequence hasn't been reached, the cursor rewinds to the last indexed snapshot.
+    async fn update(&mut self, logs: Vec<(T, LogMeta)>, range: RangeInclusive<u32>) -> Result<()> {
+        // Remove any duplicates, filter out any logs preceding our current snapshot, and sort in ascending order.
+        let logs = logs
+            .into_iter()
+            .dedup_by(|(log_a, _), (log_b, _)| log_a.sequence() == log_b.sequence())
+            .filter(|(log, _)| log.sequence() >= self.current_indexing_snapshot.sequence)
+            .sorted_by(|(log_a, _), (log_b, _)| log_a.sequence().cmp(&log_b.sequence()))
+            .collect::<Vec<_>>();
+
+        let all_log_sequences = logs
+            .iter()
+            .map(|(log, _)| log.sequence())
+            .collect::<HashSet<_>>();
+
+        match &self.index_mode {
+            IndexMode::Sequence => {
+                // We require that we've gotten all sequences in the range.
+                let expected_sequences = range.clone().collect::<HashSet<_>>();
+                if all_log_sequences != expected_sequences {
```

Comment:
I'm realizing the range checking could be more forgiving - e.g. if the expected range is 100..=104 and we get logs with sequences [100, 101, 103, 104] (missing 102), as it's written now it'll throw out the range and rewind to try looking for 100 again. Instead it could accept 100 and 101 and then start looking for 102.

I'm open to making a change to have this be considered - however because we will have still written all the logs to the log store, and because we have logic to fast-forward, we end up having pretty much that same behavior. On the next call to `next_action`, it'll fast forward past 100 and 101 because it'll see that it's already in the db, and then start looking for 102